docs/architecture-decisions/0004-use-arrays-for-mapping-schema.md

```diff
@@ -0,0 +1,58 @@
+# 4. Use arrays for mapping schema
+
+Date: 2018-09-02
+
+## Status
+
+Accepted
+
+## Context
+
+The Java implementation of the protocol uses `final static` properties to define
+the schema for requests and responses. These are defined using objects, making
+it easy to support multiple versions for each API call.
+
+PHP doesn't have this feature so we need decide how to solve this.
+
+## Decision
+
+We're still going to use objects to manipulate data and write/read content
+to/from Kafka, however it would be simpler to use arrays in constants of each
+request/response class.
+
+Each field would be an array item, the key would be the field name and the value
+would be the field type (or another array for more complex configuration).
+
+The mapping would like this:
+
+```php
+use Lcobucci\Kafka\Protocol\Type;
+
+final class DoSomethingRequest
+{
+    private const SCHEMAS = [
+        [
+            'error_code'   => Type\Int16::class,
```

Comment:
How is an optional element expressed, let's say `error_code`. Would you then need to give the configuration options as an array (like with `api_versions` below)? It's a bit confusing to me now because `api_versions` is both an array type (so would automatically have an array as it's content) _and_ is nullable.

Reply:
ah, so that's what I didn't get: some types _are_ others _aren't_ nullable. Thanks for the explanation 😎

Reply:
as they _should!_ Death to `null`!!!!